drivers/entropy/entropy_ll_nrfxlib.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <entropy.h>
+#include <soc.h>
+#include <device.h>
+#include <irq.h>
+#include <kernel.h>
+#include <kernel_includes.h>
+#include <ble_controller.h>
+#include <ble_controller_soc.h>
+
+#include <multithreading_lock.h>
+
+#define DEV_DATA(dev) \
+	((struct rng_driver_data *)(dev)->driver_data)
+
+/* The RNG driver data. */
+struct rng_driver_data {
+	/* Used to fill the pending client buffer with new random values after
+	 * RNG IRQ is triggered.
+	 */
+	struct k_sem sem_sync;
+};
+
+static struct rng_driver_data rng_data;
+
+static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+
+	u8_t *p_dst = buf;
+	u16_t bytes_left = len;
+
+	while (bytes_left > 0) {
+		s32_t bytes_read = 0;
+
+		while (bytes_read <= 0) {
+			s32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+			if (errcode) {
+				return errcode;
+			}
+			bytes_read = ble_controller_rand_vector_get(p_dst,
+								    bytes_left);
+			MULTITHREADING_LOCK_RELEASE();
+
+			if (!bytes_read) {
+				/* Put the thread on wait until next interrupt
+				 * to get more random values.
+				 */
+				k_sem_take(&rng_data.sem_sync, K_FOREVER);
+			}
+		}
+
+		p_dst += bytes_read;
+		bytes_left -= bytes_read;
+	}
+
+	return 0;
+}
+
+static int rng_driver_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len,
+				      u32_t flags)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+	s32_t errcode;
+
+	if (likely((flags & ENTROPY_BUSYWAIT) == 0)) {
+		errcode = MULTITHREADING_LOCK_ACQUIRE_NO_WAIT();
+		if (!errcode) {
+			errcode = ble_controller_rand_vector_get(buf, len);
+			MULTITHREADING_LOCK_RELEASE();
+		}
+		return errcode;
+	}
+
+	errcode = MULTITHREADING_LOCK_ACQUIRE_FOREVER_WAIT();
```

Comment:
`...LOCK_ACQUIRE_NO_WAIT` was ok (if implemented right) but here is `...LOCK_ACQUIRE_FOREVER_WAIT`

Reply:
.. hm, alright, so use NO_WAIT for both and return -error if we cannot lock or if there are not enough random bytes available.

Btw, isn't the entropy driver in Zephyr effectively blocking the calling thread as well?
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/entropy/entropy_nrf5.c#L279